src/security.js

```diff
@@ -0,0 +1,326 @@
+      });
+  }
+
+  /**
+   * Authenticates that the request is allowed to continue based on
+   * the authentication requirements
+   * @param {AuthenticationRequirement} [authentication] the requirements
+   *    for authentication to succeed
+   * @param {Request} req the request
+   * @returns {Promise<void>}
+   */
+  async authenticate(req, authentication) {
+    const token = this.#getToken(req);
+    const spaceId = this.#getRequiredHeader(req, 'x-space-id');
+    const payload = jwt.decode(token);
+
+    // first validate that the tenant is correct
+    if (payload.tenantId !== spaceId) {
+      this.#log.debug(
+        `Mismatched spaceId, expected ${spaceId}, found ${payload.tenantId}`,
+      );
+      throw new RestError(403);
+    }
+    const { allowedRoles } = authentication || {};
+    if (allowedRoles && allowedRoles.length > 0) {
+      const actualRoles = payload.roles || [];
+      if (!allowedRoles.some((role) => actualRoles.includes(role))) {
+        this.#log.debug('Payload did not contain allowed roles', {
+          allowedRoles: authentication.allowedRoles,
+          actualRoles,
+        });
+        throw new RestError(403);
+      }
+    }
+
+    const { allowedPermissions } = authentication || {};
+    if (allowedPermissions && allowedPermissions.length > 0) {
+      const permissions = payload.permissions || [];
+      if (!Security.#hasPermissions(allowedPermissions, permissions)) {
+        this.#log.debug('Payload did not contain allowed permissions', {
+          allowedRoles: authentication.allowedPermissions,
+          roles: payload.roles,
+        });
+        throw new RestError(403);
+      }
+    }
+  }
+
+  /**
+   * Verifies that the request has an authorization header and that the value
+   * of that header is a valid JWT token signed with the Private key associated
+   * with the provided Public key.
+   * @param {Request} req the request to authorize
+   * @returns {Promise<void>} resolves if the request is authorized
+   */
+  async authorize(req) {
+    this.#getRequiredHeader(req, 'x-space-id');
+    const token = this.#getToken(req);
+    await this.#verifyToken(token);
+  }
+
+  /**
+   * Generates a server to server token with the specified settings
+   * @param {TokenRequest} request
+   * @returns {Promise<string>}
+   */
+  async generateToken(request) {
+    const authToken = await this.#getAuthToken();
+    if (!this.#roles) {
+      this.#roles = await this.#getRoles(authToken);
+    }
+    const roleIds = request.roleKeys.map((k) => this.#roles[k]);
+
+    return this.#generateAccessToken(
+      authToken,
+      request.spaceId,
+      request.generator,
+      roleIds,
+      request.expiresInMinutes || 20160,
+    );
+  }
+
+  /**
+   *
+   * @param {string} authToken
+   * @param {string} spaceId
+   * @param {string} generator
+   * @param {Array<string>} roleIds
+   * @param {number | undefined} expiresInMinutes
+   */
+  async #generateAccessToken(
```

Comment:
in the process through the UI -> backend, the request will already have this access token from frontegg right?

so token generation is only needed in situations where the endpoint is called from outside frontegg login screen (ie extractors or manual requests)

Reply:
Right now, the only real case right now is to support the IT, however, this should be used for backend calls (for example Dropbox Extractor calling Ingestor)